internal/daemon/vendor-specific-plugins/common/vspnetutils.go

```diff
@@ -110,3 +132,134 @@ func GetNetDevNameFromPCIeAddr(platform platform.Platform, pcieAddress string) (
 
 	return "", fmt.Errorf("network device not found for PCI address %s", pcieAddress)
 }
+
+// CreateVethPair function to create a veth pair with the given index and InterfaceInfo
+func CreateVethPair(index int) (*VEthPairDeviceInfo, error) {
+	//nfInterfaceName is the name of the interface on the Network Function attached to the container
+	//dpInterfaceName is the name of the interface on the Data Plane OvS side
+	nfInterfaceName := fmt.Sprintf("nf_if%d", index)
+	dpInterfaceName := fmt.Sprintf("dp_if%d", index)
+
+	var nfLink netlink.Link
+	var dpLink netlink.Link
+	var err error
+
+	// Create early for clean up purposes.
+	deviceInfo := VEthPairDeviceInfo{
+		NfInterfaceName: nfInterfaceName,
+		DpInterfaceName: dpInterfaceName,
+		DpMAC:           "",
+	}
+
+	// Destroy existing veth pair if it exists
+	// This is to ensure that we do not have any stale veth pairs lying around.
+	// Error is fine if there is no existing veth pair.
+	_ = DestroyVethPair(&deviceInfo)
+
+	nfLink, _ = netlink.LinkByName(nfInterfaceName)
+	if nfLink != nil {
+		dpLink, _ = netlink.LinkByName(dpInterfaceName)
+	}
+
+	if nfLink == nil || dpLink == nil {
+		vethLink := &netlink.Veth{
+			LinkAttrs: netlink.LinkAttrs{Name: nfInterfaceName},
+			PeerName:  dpInterfaceName,
+		}
+		if err := netlink.LinkAdd(vethLink); err != nil {
+			klog.Errorf("Error occurred in creating veth pair: %v", err)
+			return nil, err
+		}
+
+		nfLink, err = netlink.LinkByName(nfInterfaceName)
+		if err != nil {
+			klog.Errorf("Error occurred in getting Link By Name of NF Link: %v", err)
+			return nil, err
+		}
+		dpLink, err = netlink.LinkByName(dpInterfaceName)
+		if err != nil {
+			klog.Errorf("Error occurred in getting Link By Name of DP Link: %v", err)
+			return nil, err
+		}
+	}
+
+	if err := netlink.LinkSetUp(nfLink); err != nil {
+		klog.Errorf("Error occurred in setting up NF link: %v", err)
+		errOnDestroy := DestroyVethPair(&deviceInfo)
+		if errOnDestroy != nil {
+			klog.Errorf("Error occurred in destroying existing veth pair: %v", err)
+			err = errors.Join(errOnDestroy, err)
+		}
+		return nil, err
+	}
+
+	if err := netlink.LinkSetUp(dpLink); err != nil {
+		klog.Errorf("Error occurred in setting up DP link: %v", err)
+		errOnDestroy := DestroyVethPair(&deviceInfo)
+		if errOnDestroy != nil {
+			klog.Errorf("Error occurred in destroying existing veth pair: %v", err)
+			err = errors.Join(errOnDestroy, err)
+		}
+		return nil, err
+	}
+
+	deviceInfo.DpMAC = dpLink.Attrs().HardwareAddr.String()
+
+	return &deviceInfo, nil
```

Comment:
as a general purpose function, `CreateVethPair()` should return two `netlink.Link` instances (or an error).

If some upper layer wants to put some parts of that into a `deviceInfo` tuple, they can do so themselves. What I mean is that (general purpose) `CreateVethPair` should not return a less useful `deviceInfo`, but rather the netlink.Link that it already has. And any caller can do with them what they want.

Reply:
netlink.Link can become stale  I believe. It is better that the callers derive it from a fresh netlink call.